src/compactor_state.rs

```diff
@@ -138,11 +137,62 @@ impl CompactorState {
                 return Err(SlateDBError::InvalidCompaction);
             }
         }
+
+        self.validate_compaction(&compaction.sources, compaction.destination)?;
+
         info!("accepted submitted compaction: {:?}", compaction);
         self.compactions.insert(compaction.destination, compaction);
         Ok(())
     }
 
+    pub(crate) fn validate_compaction(
+        &mut self,
+        sources: &[SourceId],
+        destination: u32,
+    ) -> Result<(), SlateDBError> {
+        let compacted = &self.db_state.compacted;
+        let sources_logical_order: Vec<SourceId> = self
+            .db_state
+            .l0
+            .iter()
+            .map(|sst| SourceId::Sst(sst.id.unwrap_compacted_id()))
+            .chain(compacted.iter().map(|sr| SourceId::SortedRun(sr.id)))
+            .collect();
+
+        let mut start_idx = 0;
+        while start_idx < sources_logical_order.len() && sources_logical_order[start_idx] != sources[0] {
+            start_idx += 1;
+        }
+
+        if start_idx + sources.len() > sources_logical_order.len() ||
+            !sources.iter().eq(&sources_logical_order[start_idx..start_idx + sources.len()])
+        {
+            return Err(SlateDBError::InvalidCompaction);
+        }
+
+        match sources.last() {
+            Some(SourceId::Sst(_)) => {
+                if let Some(first_sr) = compacted.first() {
+                    if destination < first_sr.id {
+                        return Err(SlateDBError::InvalidCompaction);
+                    }
+                }
+            }
+            Some(SourceId::SortedRun(last_sr)) => {
+                if compacted.iter().any(|sr| sr.id == destination) {
+                    if destination != *last_sr {
+                        return Err(SlateDBError::InvalidCompaction);
+                    }
+                } else {
+                    return Err(SlateDBError::InvalidCompaction);
```

Comment:
can you add an explanatory log line like `warn!("sorted run specifies a sorted run as the last source and destination sorted run is not present in current compacted list")`